modules/charts/src/components/Provider/Provider.tsx

```diff
@@ -0,0 +1,119 @@
+	getChartData: ({ fieldName }: { fieldName: string }) => {
+		isLoading: boolean;
+		isError: boolean;
+		// TODO: Map<string, Aggregtions | NumericAggregations>
+		data: Map<string, {}> | undefined;
+	};
+};
+
+export const ChartsContext = createContext<ChartContextType | null>(null);
+
+type ChartsProviderProps = React.PropsWithChildren<{
+	// ChartProviderTheme vs ChartTheme (global chart vs individual chart)
+	// viz vs chart, say "chart" everywhere, differentiate between the atual viz and the components
+	theme: { vizColors: {} };
+	components?: {
+		Tooltip: ReactElement;
+		Error: ReactElement;
+		Loader: ReactElement;
+	};
+}>;
+
+const createChartDataMap = ({ data }) => {
+	if (!data) {
+		return null;
+	}
+	// TODO: Dynamic property
+	return new Map(Object.entries(data.data.file.aggregations));
+};
+
+export const ChartsProvider = ({
+	theme = { vizColors: ['red', 'green', 'blue'] },
+	components,
+	children,
+}: ChartsProviderProps) => {
+	console.log('CHARTS PROVIDER');
+	const isInitialized = useRef(false);
+
+	// call once, not a hook
+	const { resolveColor, createColorMap } = useMemo(() => createChartColors({ colors: theme.vizColors }), []);
+
+	// apiFetcher is consumer function passed into ArrangerDataProvider, currently no default
+	const { documentType, apiFetcher, sqon, setSQON } = useArrangerData({
+		callerName: 'ArrangerCharts',
+	});
+
+	const { apiState, addToQuery, removeFromQuery } = useNetworkQuery({
+		documentType,
+		apiFetcher,
+		sqon,
+	});
+
+	// first time? maybe hook for state, closer to derived state
+	const chartDataMap = createChartDataMap({ data: apiState?.data });
+
+	if (
+		apiState.loading === false &&
+		apiState.error === false &&
+		apiState.data !== undefined &&
+		isInitialized.current === false &&
+		chartDataMap
+	) {
+		/**
+		 * need consistent keys even when query changes and becomes filtered down
+		 * initialize once with all keys
+		 */
+		createColorMap(chartDataMap);
+		isInitialized.current = true;
+	}
+
+	const chartContext: ChartContextType = {
+		theme,
+		registerChart: async ({ fieldName }) => {
+			addToQuery({ fieldName });
+		},
+
+		deregisterChart: ({ fieldName }) => {
+			removeFromQuery({ fieldName });
+		},
+
+		update: ({ fieldName, eventData }) => {
+			console.log('update', fieldName, eventData);
+			// new data => sqon => arranger => data => render
+			// update arranger.setSqon
+			setSQON();
+		},
+
+		// chartType for slicing data
+		getChartData: ({ fieldName }) => {
+			const chartData = chartDataMap?.get(fieldName);
+
+			return {
+				isLoading: false,
+				isError: false,
+				data: chartData,
+			};
+		},
+		resolveColor,
+	};
+
+	return <ChartsContext.Provider value={chartContext}>{children}</ChartsContext.Provider>;
+};
+
+export const useChartsContext = () => {
+	const context = useContext(ChartsContext);
+	if (!context) {
+		throw new Error('context has to be used within <Charts.Provider>');
```

Comment:
this may lead to unhandled error situations as it relies on the consumer side accounting for the lack of a "default" fallback behaviour. e.g. provide an empty context if the expected one is unavailable.

... probably feedback also applicable to the OG Arranger UI components

Reply:
I think yelling with a clear error on the consumer side is better than "silent failure" from a context object that doesn't make any sense.

I believe this is a similar context check to `useArrangerData` and `ArrangerDataProvider`

Reply:
For sure! I'd like your input on the best way to handle it